mypy/newsemanal/semanal.py

```diff
@@ -1856,16 +1859,42 @@ def add_type_alias_deps(self, aliases_used: Iterable[str],
             target = self.scope.current_target()
         self.cur_mod_node.alias_deps[target].update(aliases_used)
 
+    def is_not_ready_alias(self, rv: Expression) -> bool:
+        """Does this expression look like a not ready type alias?
+
+        This includes 'Ref' and 'Ref[Arg1, Arg2, ...]', where 'Ref'
+        refers to a PlaceholderNode with becomes_typeinfo=True.
+        """
+        if isinstance(rv, IndexExpr) and isinstance(rv.base, RefExpr):
+            return self.is_not_ready_alias(rv.base)
+        if isinstance(rv, NameExpr):
+            n = self.lookup(rv.name, rv)
+            if n and isinstance(n.node, PlaceholderNode) and n.node.becomes_typeinfo:
+                return True
+        elif isinstance(rv, MemberExpr):
+            fname = get_member_expr_fullname(rv)
+            if fname:
+                n = self.lookup_qualified(fname, rv)
+                if n and isinstance(n.node, PlaceholderNode) and n.node.becomes_typeinfo:
+                    return True
+        return False
+
     def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
         s.is_final_def = self.unwrap_final(s)
         tag = self.track_incomplete_refs()
         s.rvalue.accept(self)
-        if self.found_incomplete_ref(tag):
+        if isinstance(s.rvalue, IndexExpr) and isinstance(s.rvalue.base, RefExpr):
+            # Special case: analyze index expression _as a type_ to trigger
+            # incomplete refs for string forward references, for example
+            # Union['ClassA', 'ClassB']
+            self.analyze_alias(s.rvalue, allow_placeholder=True)
+        is_alias = self.is_not_ready_alias(s.rvalue)
+        if self.found_incomplete_ref(tag) or is_alias:
             # Initializer couldn't be fully analyzed. Defer the current node and give up.
             # Make sure that if we skip the definition of some local names, they can't be
             # added later in this scope, since an earlier definition should take precedence.
             for expr in names_modified_by_assignment(s):
-                self.mark_incomplete(expr.name, expr)
+                self.mark_incomplete(expr.name, expr, becomes_typeinfo=is_alias)
```

Comment:
What if the alias is like `X = List[Y]` or `X = C[Y]` where only `Y` is not ready? It's not clear to me how we'd set `becomes_typeinfo` correctly in that case.

Reply:
In this case we can turn them into `Var`s, this is what I mentioned in private chat. In general, I don't like the current situation, I will try a bigger change, maybe there is a way to make it more straightforward.